csa/man/csacrep.8.in
.\"
.\" (C) 2000-2007 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This documentation supports a
.\" program that is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" The documentation that supports
.\" this program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this documentation and the software it
.\" supports, are distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\"
.\" Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
.\" Sunnyvale, CA  94085, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan
.\"
.TH csacrep 8
.SH NAME
csacrep \- reports on consolidated accounting data
.SH SYNOPSIS
\f4@sbindir@/csacrep\f1
[\f4-p\f1\ |\ \f4-u\f1]
[\f4-b\f1]
[\f4-c\f1]
[\f4-d\f1]
[\f4-f\f1]
[\f4-g\f1]
[\f4-h\f1]
[\f4-j\f1]
[\f4-n\f1]
[\f4-w\f1]
[\f4-x\f1]
[\f4-J\f1]
[\f4-P\f1]
[\f4-V\f1]
.SH DESCRIPTION
The \f4csacrep\f1 command generates reports
from data in \f4cacct\f1 format, such as output from
the \f4csacon\f1(8) command.
The default output reports the project name, the user ID,
and the login name, and it suppresses the heading.
.PP
The \f4csacrep\f1 command accepts the following two types of options:
sorting and printing.
.SS Sorting Options
The \f4csacrep\f1 command accepts the following two sorting options.
You cannot use the \f4-p\f1 and \f4-u\f1 options together.
The \f4-p\f1 option is currently deferred, consequently, the sort
currently is only by user ID.
.TP 5
\f4-p\f1
Sorts output first by the project ID (deferred) 
and then by the secondary key of the user ID.
By default, \f4csacrep\f1 does not sort.
You cannot use this option with the \f4-u\f1 option.
.TP
\f4-u\f1
Sorts output first by user ID and then by the secondary key of the project ID (deferred).
By default, \f4csacrep\f1 does not sort.
You cannot use this option with the \f4-p\f1 option.
.SS Printing Options
The \f4csacrep\f1 command accepts the following printing options:
.TP 5
\f4-b\f1
Reports SBU data.
.TP
\f4-c\f1
Reports CPU time, total kcore-minutes, and total kvirtual-minutes.
.TP
\f4-d\f1
Reports the cumulative disk usage and number of disk samples.
Input files that contains disk usage data are generated by using
\f4acctdisk\f1 (see acct(8)) or by merging \f4cacct\f1 disk output files
with other \f4cacct\f1 files, using the \f4csaaddc\f1(8) command.
.TP
\f4-f\f1
Reports full data.
.TP
\f4-g\f1
Reports group name.
.TP
\f4-h\f1
Displays heading.
.TP
\f4-j\f1
Reports number of processes and jobs.
.TP
\f4-n\f1
Reports prime and nonprime data.
.TP
\f4-w\f1
Reports the CPU, block I/O, and swap in delays.
.TP
\f4-x\f1
Reports the number of k-characters read and written, 
and the number of read and write systems calls.
.TP
\f4-J\f1
Reports job ID.
.TP
\f4-P\f1
Prints the project ID. (deferred)
(The default is to print the project name.)
.TP
\f4-V\f1
Prints the number of minor and major page faults.
.SH NOTES
Zero is a valid value for the number of jobs.
If a job is executed with multiple user ID/project ID pairs,
the number-of-jobs value is incremented for only one such
combination per job.
.SH EXAMPLES
The following example generates a report from a daily accounting file:
.RS .5i
.PP
\f4csacrep -hcw < @csalocalstatedir@/sum/cacct.02031315\f1
.RE
.PP
.SH SEE ALSO
csaaddc(8),
csacon(8).


csa/ChangeLog
* Wed Jun  8 2016 Jay Lan <jay.j.lan@nasa.gov>
- Converted csa package from the last SGI-supported version 4.1.1 to
  community version 5.0.0

* Fri Oct 17 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.1
- Various fixes to allow for SLES11 build

* Fri Jul 25 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.0
- Fixed memory leaks
- Changed acctjob structure offset type from int to int64_t

* Fri Apr 18 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.2
- Updated to use new libnl API
- Fixed process exit code assignment

* Tue Mar 4 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.1
- Cleaned up acctdusg
- Added check to prevent SIGPIPE when communicating with daemon
- Increased maximum PID allowed for larger systems

* Wed Nov 7 2007 Jonathan Lim <jlim@sgi.com>
- csa-4.0.0
- Implement kernel functionality as a userspace daemon
- Remove obsolete wait metrics and replace with taskstats delay accounting
- Various bug fixes

* Fri Apr 27 2007 Jay Lan <jlan@sgi.com>
* Mon Apr 30 2007 Jonathan Lim <jlim@sgi.com>
- csa-3.0.2
- SBU's data was all zero in csa-3.0.0 and csa-3.0.1
- Clarify csacom(1) man page on '-f' option
- Obsolete csacom '-a' & '-q' option
- Fix a problem of csadrep reading input from a binary file
- Fixed csadrep(8) man page: replaced '-w' with '-n' and removed '-t'

* Wed Jul 5 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.1
- 'csacom -b' was broken in csa-3.0.0
- 'csacom -b' drops the last record of a pacct file

* Thu Feb 24 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.0
- csacom -[esES] option was broken.
- create /usr/sbin/rccsa symlink at %install in the spec file
- remove code for CSA record types not supported in Linux
- spliting up csa_kern.h into kernel version and userland version
- support pacct files created in csa-2.*.*

* Wed Jul 27 2005 Chris Sturtivant <csturtiv@sgi.com>
- csa-2.2.2
- Disabled perror for permission denied in csa_ctl when called by csa_auth

* Mon May 2 2005 Jay Lan <jlan@sgi.com>
- csa-2.2.1
- clean up csa init script
- csa_wracct(3) man page update
- 'csacom -o' wrote only one record
- 'csacom -b' does not work
- csajrep man page references invalid option '-W'
- 'csajrep -r' does not work
- 'dodisk' created files with incorrect user and group name 'adm'
- CSA does not handle gid's > 65535
- 'csacom -j' failed if job id greater than max value of an int

* Mon Nov 22 2004 Jay Lan <jlan@sgi.com>
- csa-2.2.0
- Fixed compilation issue with gcc 3.3.3
- CSA assumes existence of 'adm' user/group account
- csacom & csacms usage otuput missing carriage return
- csarecy misses a carriage return on certain condition
- csacom -W segfaults after printing one bad time

* Fri Oct 8 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.2
- Fixed 'csaswitch -c halt' bug
- Fixed 'csaedit dumped core on bad wkmg record' bug

* Tue Jul 27 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.1
- Remove static API library libcsa.a from rpm package
- Fixed a bug at csacom with invalid -u option
- cascom headers and data not aligned correctly at output

* Wed Jun 30 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.0
- Fixed a problem dealing with accounting files > 2GB
- Now derives "Day of Year" from other fields of the holidays file
- Removed "Pages Swapped" from csacom and csacms - not supported in Linux
- Fixed a bug of not mapping UID 0 to root

* Tue Feb 17 2004 Jay Lan <jlan@sgi.com>
- csa-2.0.0
- Implemented /proc ioctl interface
- Renamed rpm from csa-job-acct to csa
- Renamed /etc/init.d/csaacct to /etc/init.d/csa

* Fri Nov 7 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-2.0.0
- Provided API library libcsa.so and associated header file

* Wed Sep 3 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-1.0.3
- Updated the spec file

* Thu Feb 13 2003 Robin Holt <holt@sgi.com>
- Updated init scripts to use /var/lock/subsys

* Tue Nov 5 2002 Richard Offer <offer@sgi.com>
- Fixed packager/distribution/vendor

* Thu May 23 2002 Robin Holt <holt@sgi.com>
- csa-job-acct-1.0.2
- Changed install destination from /usr/local to /usr
- Updated the man page location to /usr/share/man
- Added a clean and realclean target to the Makefile

* Thu Apr 5 2001 Marlys Kohnke <kohnke@sgi.com>
- Changed csadrep internal markers back to match those in IRIX CSA
- Changed ja summary report so that all lines are displayed even if the
  resource used values are zero

* Sun Feb 11 2001 Marlys Kohnke <kohnke@sgi.com>
- Added support for Workload Management daemon accounting records
- Fixed csachkpacct script problem (df -k needs to be df -kP; changed awk
  field to 4)

* Fri Jan 19 2001 Daryl Grunau <dwg@acl.lanl.gov>
- port to Linux from SGI IRIX
- csa-job-acct-1.0.0 (available January 19, 2001)
- initial csa-job-acct-1.0.0 open source release


csa/cmd/csacrep.c
/*
 * Copyright (c) 2000-2002 Silicon Graphics, Inc and LANL  All Rights Reserved.
 * Copyright (c) 2004-2007 Silicon Graphics, Inc All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as 
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it would be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 *
 * You should have received a copy of the GNU General Public License along 
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
 * Sunnyvale, CA  94085, or:
 * 
 * http://www.sgi.com 
 */

/*
 * csacrep:
 *	Output a report from a consolidated accounting file (cacct).
 *	
 * Options include:
 *	-b	Output fee and SBU data
 *	-c	Output CPU time, core and virtual memory integrals
 *	-d	Output cumulative disk usage count and # of samples
 *	-f	Report full data
 *	-g	Output group ID
 *	-h	Display heading
 *	-J	Output job ID
 *	-j	Output # of procs and jobs
 *	-n	Output prime/non-prime data
 *	-P	Print project number (default is project name)
 *	-p	Sort by project ID (default is no sort)
 *	-V	Output minor and major faults
 *	-u	Sort by user ID (default is no sort)
 *	-w	Output CPU, block I/O, and swap in delays
 *	-x	Output characters transferred, # of
 *	  	logical requests (read and write system calls)
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif	/* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#include <sys/types.h>
#include <sys/param.h>
#include <sys/sysmacros.h>

#include <string.h>
#include <time.h>
#include <unistd.h>

#include "acctdef.h"
#include "acctmsg.h"
#include "cacct.h"
#include "sbu.h"

#define	ISIZ     200      /* initial number of table entries */
#define	ZERO_MT 0.000001  /* CPU time floor for multi-tasking record */

char  *Prg_name;

int    MTASK_HI;          /* high mtask report bucket */
int    MTASK_LO;          /* low mtask report bucket */

int    b_flag = 0;
int    c_flag = 0;
int    d_flag = 0;
int    f_flag = 0;
int    g_flag = 0;
int    h_flag = 0;
int    J_flag = 0;
int    j_flag = 0;
int    n_flag = 0;
int    P_flag = 0;
int    V_flag = 0;
int    w_flag = 0;
int    x_flag = 0;

int    Psort = 0;
int    Usort = 0;

struct cacct  *clist;
int    csize = 0;         /* current entries used */
int    max_size = 0;      /* max table size */

char  *optstring = "bcdfghJjnPpVuwx";

void enter(struct cacct *);
void full_print(struct cacct *);
void hdprint();
void select_print(struct cacct *, int);

void usage()
{
    acct_err(ACCT_ABORT,
           "Command Usage:\n%s\t[-p | -u] [-bcdfghJjnPVwx]",
    	Prg_name);
}

cmp(const void *e1, const void *e2)
{
    struct cacct  *c1;
    struct cacct  *c2;

    c1 = (struct cacct *) e1;
    c2 = (struct cacct *) e2;

    if (Psort)
    {
        if (c1->ca_prid - c2->ca_prid)
            return(c1->ca_prid - c2->ca_prid);
        else
            return(c1->ca_uid - c2->ca_uid);
    }
    else
    {
        if (c1->ca_uid - c2->ca_uid)
            return(c1->ca_uid - c2->ca_uid);
        else
            return(c1->ca_prid - c2->ca_prid);
    }
}

main(int argc, char **argv) 
{
    int           c;
    struct cacct  cbuf;
    int	          fd;
    int           one = 1;

    Prg_name = argv[0];
    while ((c = getopt(argc, argv, optstring)) != EOF)
    {
        switch((char) c)
        {
        case 'b':
            b_flag++;
            break;
        case 'c':
            c_flag++;
            break;
        case 'd':
            d_flag++;
            break;
        case 'f':
            f_flag++;
            break;
        case 'g':
            g_flag++;
            break;
        case 'h':
            h_flag++;
            break;
        case 'J':
            J_flag++;
            break;
        case 'j':
            j_flag++;
            break;
        case 'n':
            n_flag++;
            break;
        case 'P':
            P_flag++;
            break;
        case 'p':
            Psort = 1;
            break;
	case 'V':
	    V_flag = 1;
	    break;
        case 'u':
            Usort = 1;
            break;
        case 'w':
            w_flag++;
            break;
        case 'x':
            x_flag++;
            break;
        default:
            usage();
            exit(1);
        }
    }

    if (Psort + Usort > 1)
    {
        acct_err(ACCT_FATAL,
               _("The (-%s) and the (-%s) options are mutually exclusive."),
        	"p", "u");
        usage();
        exit(1);
    }

    /*
     *	Get the system parameters from the configuration file.
     */
    if ((fd = openacct (NULL, "r")) < 0)
        acct_perr(ACCT_ABORT, errno,
        	_("An error occurred during the opening of file '%s'."),
        	"stdin");
    if (Psort + Usort == 0)
    {
        /*
         *	Read each cacct record and write the requested output.
         */
        while (readcacct (fd, &cbuf) == sizeof (struct cacct))
        {
            if (f_flag)
                full_print(&cbuf);
            else
            {
                if (h_flag && one)
                {
                    one = 0;
                    hdprint();
                }
                select_print(&cbuf, 0);
            }
        }
    }
    else
    {
        int  i;
        
        while (1)
        {
            if (readcacct (fd, &cbuf) == sizeof (struct cacct))
                enter(&cbuf);
            else
                break;
        }
        qsort((char *) clist, csize, sizeof(struct cacct), cmp);
        for (i = 0; i < csize; i++)
        {
            if (f_flag)
                full_print(&clist[i]);
            else
            {
                if (h_flag && one)
                {
                    one = 0;
                    hdprint();
                }
                select_print(&clist[i], 0);
            }
        }
    }

    exit(0);
}

void enter(struct cacct *rec)
{
    int  sz;

    /*
     *	If first call.
     */
    if (max_size == 0)
    {
        max_size = ISIZ;
        sz = sizeof(struct cacct) * max_size;
        if ((clist = (struct cacct *) malloc(sz)) == NULL)
            acct_perr(ACCT_ABORT, errno,
            	_("There was insufficient memory available when allocating '%s'."),
            	"struct cacct");
    }

    /*
     *	If table is overflowing.
     */
    if (csize == max_size)
    {
        max_size = max_size + ISIZ;
        sz = sizeof(struct cacct) * max_size;
        if ((clist = (struct cacct *) realloc(clist, sz)) == NULL)
            acct_perr(ACCT_ABORT, errno,
            	_("There was insufficient memory available when reallocating '%s'."),
            	"struct cacct");
    }

    /*
     *	Add to table.  Let compiler do vector copy.
     */
    clist[csize++] = *rec;
}

/*
 *	Full print routine.
 */
void full_print (struct cacct *rec)
{
    int    dm_used = 0;
    char  *group;
    int    i;
    char  *name;
    char  *project;
    int    tape_used = 0;
    char  *user;

    project = prid_to_name(rec->ca_prid);
    if (project[0] == '?')
        project = "Unknown";
    printf ("Project Name / PRID : %8.8s / %-5lld  ", project, rec->ca_prid);

    user = uid_to_name(rec->ca_uid);
    if (user[0] == '?')
        user = "Unknown";
    printf ("Login Name / UID   : %8.8s / ", user);
    if (rec->ca_uid == CACCT_NO_ID)
        printf ("Unknown\n");
    else
        printf ("%d\n", rec->ca_uid);
    
    printf ("No. of Processes   : %8lld           ", rec->ca_pc);
    printf ("No. of Jobs        : %8lld\n", rec->ca_njobs);
    printf ("Disk Blocks        : %8.0f           ", rec->ca_du);
    printf ("Disk Samples       : %8d\n", rec->ca_dc);
    
    printf ("Job Id             : ");
    if (rec->ca_jid == CACCT_NO_ID)
        printf ("%8.8s           ", "Unknown");
    else
        printf ("%-#18llx ", rec->ca_jid);
    printf ("Group Id           : ");
    group = gid_to_name(rec->ca_gid);
    if (group[0] == '?')
        group = " Unknown";
    printf ("%8.8s\n", group);
    
    printf ("Fee                : %8.2f           ", rec->ca_fee);
    printf ("SBU's              : %8.2f\n", rec->ca_sbu);
    /* printf ("Migrated Blocks    : %8lld\n", rec->ca_dmdu); */
    printf ("                              --------------------\n");

    printf ("CPU time               (P/N) : %8.0f / %8.0f  seconds\n",
            rec->prime.cpuTime, rec->nonprime.cpuTime);
    printf ("KCORE * CPU Time       (P/N) : %8.0f / %8.0f  Kbytes * mins\n",
            rec->prime.kcoreTime, rec->nonprime.kcoreTime);
    printf ("KVIRT * CPU Time       (P/N) : %8.0f / %8.0f  Kbytes * mins\n",
            rec->prime.kvirtualTime, rec->nonprime.kvirtualTime);
    printf ("Total K-Chars Xferred  (P/N) : %8.2f / %8.2f \n",
            rec->prime.charsXfr / 1024.0, rec->nonprime.charsXfr / 1024.0);
    printf ("Total Log. I/O Req.    (P/N) : %8lld / %8lld \n",
            rec->prime.logicalIoReqs, rec->nonprime.logicalIoReqs);
    printf ("Total Minor Faults     (P/N) : %8lld / %8lld \n",
	    rec->prime.minorFaultCount, rec->nonprime.minorFaultCount);
    printf ("Total Major Faults     (P/N) : %8lld / %8lld \n",
	    rec->prime.majorFaultCount, rec->nonprime.majorFaultCount);
    printf ("CPU Delay              (P/N) : %8.0f / %8.0f  seconds\n",
            rec->prime.cpuDelayTime, rec->nonprime.cpuDelayTime);
    printf ("Block I/O Delay        (P/N) : %8.0f / %8.0f  seconds\n",
            rec->prime.blkDelayTime, rec->nonprime.blkDelayTime);
    printf ("Swap In Delay          (P/N) : %8.0f / %8.0f  seconds\n",
            rec->prime.swpDelayTime, rec->nonprime.swpDelayTime);

    /* Workload management */
    printf ("Number of Batch Requests       : %8lld\n", rec->can_nreq);
    printf ("Total User CPU Time          : %8.4f seconds\n",
            TIME_2_SECS(rec->can_utime));
    printf ("Total System CPU Time        : %8.4f seconds\n",
            TIME_2_SECS(rec->can_stime));


    /* NOTE - Workload management information needs to be printed when it is
       available. */
    

    printf ("\n");
    printf ("================================================================================\n");
    printf ("\n");
}

/*
 *	Print heading.
 */
void hdprint()
{
    int   i;
    char  t1[9];
    char  l1[4056], l2[4056], l3[4056];

    l1[0] = l2[0] = l3[0] = '\0';

    strcat (l1, "PROJECT    USER    LOGIN  ");
    if (P_flag)
        strcat (l2, "  ID        ID     NAME   ");
    else
        strcat (l2, " NAME       ID     NAME   ");
    strcat (l3, "======== ======== ========");

    if (J_flag)
    {
        strcat(l1, "        JOB        ");
        strcat(l2, "        ID         ");
        strcat(l3, " ==================");
    }

    if (g_flag)
    {
        strcat(l1, "  GROUP ");
        strcat(l2, "  NAME  ");
        strcat(l3, " =======");
    }

    if (n_flag)
    {
        strcat (l1, " PRIME/ ");
        strcat (l2, " N-PRIME");
        strcat (l3, " =======");
    }

    if (c_flag)
    {
        strcat (l1, "  CPU-TIM  KCORE *  KVIRT *");
        strcat (l2, "  [SECS]   CPU-MIN  CPU-MIN");
        strcat (l3, " ======== ======== ========");
    }

    if (w_flag)
    {
        strcat (l1, "        DELAY [SECS]       ");
        strcat (l2, "   CPU     BLOCK     SWAP  ");
        strcat (l3, " ======== ======== ========");
    }

    if (x_flag)
    {
        strcat (l1,"  K-CHARS LOG-I/O ");
        strcat (l2,"  XFERRED REQUESTS");
        strcat (l3," ======== ========");
    }

    if (V_flag)
    {
	strcat (l1,"    MINOR    MAJOR"); 
	strcat (l2,"    FAULTS   FAULTS");
	strcat (l3,"  ======== ========");
    }

    if (j_flag)
    {
        strcat (l1, "   # OF     # OF  ");
        strcat (l2, "   PROCS    JOBS  ");
        strcat (l3, " ======== ========");
    }

    if (d_flag)
    {
        strcat (l1, "   DISK     DISK  ");
        strcat (l2, "  BLOCKS   SAMPLES");
        strcat (l3, " ======== ========");
    }

    if (b_flag)
    {
        strcat (l1, "    FEE     SBUs  ");
        strcat (l2, "                  ");
        strcat (l3, " ======== ========");
    }

    printf ("%s\n", l1);
    printf ("%s\n", l2);
    printf ("%s\n", l3);
}

/*
 * pnp is assumed to be 0 when called from main().
 * pnp is 1 on recursive call if n_flag is set.
 */
void select_print (struct cacct *rec, int pnp)
{
    char  *group;
    int	   ind;
    char  *project;
    char  *user;

    if (P_flag)
        printf ("%-8lld", rec->ca_prid);
    else
    {
        project = prid_to_name(rec->ca_prid);
        if (project[0] == '?')
            project = "Unknown";
        printf ("%-8.8s", project);
    }

    user = uid_to_name(rec->ca_uid);
    if (user[0] == '?')
        user = "Unknown";
    if (rec->ca_uid == CACCT_NO_ID)
        printf (" %-8.8s", "Unknown");
    else
        printf (" %-8d", rec->ca_uid);
    printf (" %-8.8s", user);

    if (J_flag)
    {
        if (rec->ca_jid == CACCT_NO_ID)
            printf (" %18.18s", "Unknown");
        else
            printf (" %#18llx", rec->ca_jid);
    }

    if (g_flag)
    {
        group = gid_to_name(rec->ca_gid);
        if (group[0] == '?')
            group = "Unknown";
        printf (" %-7.7s", group);
    }

    if (n_flag)
        printf (" %7.7s", pnp ? "N" : "P");

    if (c_flag)
    {
        if (pnp == 0)
            printf (" %8.0f %8.0f %8.0f",
                rec->prime.cpuTime + (n_flag ? 0 : rec->nonprime.cpuTime),
                rec->prime.kcoreTime + (n_flag ? 0 : rec->nonprime.kcoreTime),
                rec->prime.kvirtualTime + (n_flag ? 0 : rec->nonprime.kvirtualTime));

	else
            printf (" %8.0f %8.0f %8.0f",
                rec->nonprime.cpuTime,
                rec->nonprime.kcoreTime,
                rec->nonprime.kvirtualTime);
    }

    if (w_flag)
    {
        if (pnp == 0)
            printf (" %8.0f %8.0f %8.0f",
                rec->prime.cpuDelayTime + (n_flag ? 0 : rec->nonprime.cpuDelayTime),
                rec->prime.blkDelayTime + (n_flag ? 0 : rec->nonprime.blkDelayTime),
		rec->prime.swpDelayTime + (n_flag ? 0 : rec->nonprime.swpDelayTime));
	else
            printf (" %8.0f %8.0f %8.0f",
                rec->nonprime.cpuDelayTime,
                rec->nonprime.blkDelayTime,
		rec->nonprime.swpDelayTime);
    }

    if (x_flag)
    {
        if (pnp == 0)
            printf (" %8.0f %8lld",
                (rec->prime.charsXfr + (n_flag ? 0 : rec->nonprime.charsXfr)) / 1024.0,
                rec->prime.logicalIoReqs + (n_flag ? 0 : rec->nonprime.logicalIoReqs));
	else
            printf (" %8.0f %8lld",
                rec->nonprime.charsXfr,
                rec->nonprime.logicalIoReqs);
    }

    if (V_flag)
    {
        if (pnp == 0)
	    printf (" %8lld %8lld",
		rec->prime.minorFaultCount + (n_flag ? 0 : rec->nonprime.minorFaultCount),
		rec->prime.majorFaultCount + (n_flag ? 0 : rec->nonprime.majorFaultCount));
	else
	    printf (" %8lld %8lld",
		rec->nonprime.minorFaultCount,
		rec->nonprime.majorFaultCount);
    }

    if (!pnp && j_flag)
        printf (" %8lld %8lld", rec->ca_pc, rec->ca_njobs);

    if (!pnp && d_flag)
        printf (" %8.0f %8d", rec->ca_du, rec->ca_dc);

    if (!pnp && b_flag)
        printf (" %8.2f %8.2f", rec->ca_fee, rec->ca_sbu);

    printf ("\n");

    if (!pnp && n_flag)
        select_print(rec, 1);

    return;
}
